### PR TITLE
Fix vote and shuffle shader instructions on AMD GPUs

### LINKED ISSUE #4104: Unreal Engine 4 games flickering on AMD GPUs
### What's the issue you encountered?
Zones flicker on the screen.

### How can the issue be reproduced?
Just start game. It happened in Densha De Go, Shin Megami Tensei V, Crisis Core - FINAL FANTASY VII– REUNION. Same on Opengl and Vulkan.

### Log file
[Ryujinx_1.1.461_2022-12-12_12-02-11.log](https://github.com/Ryujinx/Ryujinx/files/10206368/Ryujinx_1.1.461_2022-12-12_12-02-11.log)

### Environment?
- Ryujinx version: 1.1.461
- Game version: 1.0.0
- System Specs:
  - OS: Windows 11
  - CPU: R5 3600
  - GPU: AMD RX Vega 64
  - RAM: 16GiB
- Applied Mods : No

### Additional context?
https://user-images.githubusercontent.com/66036249/207003330-fc569b1b-adf3-4410-a4ed-23cd720a216d.mp4

### LINKED ISSUE #4771: [Regression] Performance loss on all games on some Intel GPUs from build 1.1.683 onwards
### Description of the issue

Build 1.1.683 #4607 has introduced a Regression to some Intel GPUs, where all games I have tested so far have suffered a large performance loss which has caused games to experience framerates from 3 to 13 on menus where they had previously experienced 30fps on previous builds. 

### Reproduction steps

- Launch any game

### Log file

[Ryujinx_1.1.682_2023-05-02_20-11-58.log](https://github.com/Ryujinx/Ryujinx/files/11374609/Ryujinx_1.1.682_2023-05-02_20-11-58.log)
[Ryujinx_1.1.743_2023-05-02_20-19-34.log](https://github.com/Ryujinx/Ryujinx/files/11374665/Ryujinx_1.1.743_2023-05-02_20-19-34.log)

### OS

Windows 11

### Ryujinx version

1.1.743

### Game version

1.0.0

### CPU

i5-1135G7

### GPU

Intel(R) Iris(R) Xe 

### RAM

16GB

### List of applied mods

_No response_

### Additional context?

1.1.682

https://user-images.githubusercontent.com/84456308/235708769-fda72bb6-a4bc-4816-b595-9aad4568355c.mp4

1.1.743

https://user-images.githubusercontent.com/84456308/235710323-a16a028a-cf01-499e-b963-4e6cdfbb6686.mp4

### LINKED ISSUE #5428: [Bug] Blinking textures (squares) all over the place in Master Detective Archives: RAIN CODE
### Description of the issue

On every level, on walls, on objects, on the floor etc. there are textures that are moving/teleporting and blinking constantly, like in the pic below

### Reproduction steps

Start the game

### Log file

[Ryujinx_1.1.952_2023-07-03_23-09-17.log](https://github.com/Ryujinx/Ryujinx/files/11940728/Ryujinx_1.1.952_2023-07-03_23-09-17.log)

### OS

Microsoft Windows 10.0.19044 (X64)

### Ryujinx version

1.1.952

### Game version

1.3.0

### CPU

Intel(R) Core(TM) i5-9400F CPU @ 2.90GHz 

### GPU

AMD Radeon RX 580 Series (Vulkan v1.3.246, Driver v2.0.262)

### RAM

16318 MiB

### List of applied mods

_No response_

### Additional context?

tG0CEaXvtlk

### PR DESCRIPTION
**Pull request by gdkchan**
Overview
---

AMD GPUs have a maximum of 64 threads per warp (or invocations per subgroup, in Khronos terms), unlike NVIDIA that has 32. Since the Switch has a NVIDIA GPU, the shaders are expecting 32 threads. In order to make them work on AMD, we need to sub-divide the 64 threads into 2 groups of 32. This change takes the extra threads into account, before this change, it just assumed that the GPU only had 32, and the other 32 were ignored, which can cause bugs.

RDNA AMD GPUs support the subgroup size control extension on Vulkan, which allows the subgroup size to be changed from 64 to 32. This partially fixes the issue, but this is only supported for compute shader, so games using subgroup operations on graphics stages would still be broken (like Marvel Ultimate Alliance 3 and Niear Automatathat uses it on geometry shaders). This change removes the use of the extension, since we can deal with the subgroup size being 64 now, there's no point in keeping it enabled, plus reducing the subgroup size can have negative performance impact on those GPUs.

Affected games
---

(They were tested on a RDNA 2 GPU with the subgroup size extension disabled, but some of them like Marvel Ultimate Alliance 3 and Nier Automata were broken before this change even with it enabled).

**Marvel Ultimate Alliance 3:**
Before:
image
After:
image
**Nier Automata: The End of YoRHa Edition:**
Before:
image
After:
image
image
**Monster Hunter Rise:**
Before:
image
image
After:
image
image
**Master Detective Archives: Rain Code:**
Before:
image
After:
image
(This one is a bit hard to notice with a static screenshot, but the "before" had squares flickering across the screen).
**Luigi's Mansion 3:**
Before:
image
After:
image
Depends on #5509 (so this is draft until that one is merged).
Fixes #4104.
Fixes #5428.
Should fix #4771 since the extension is not used anymore with this change, but I don't have that GPU to test.

Testing is welcome.

### CODE REVIEW

**Comment by Littl3Guy** [comment]
Tested with 11 games, no regressions found.

**Review by riperiperi — APPROVED** [review]
Nice to see this finally handled. Hopefully also gives Intel Xe users a break.

File: src/Ryujinx.Graphics.Shader/Instructions/InstEmitWarp.cs [hunk 1]
```diff
@@ -76,7 +63,81 @@ public static void Vote(EmitterContext context)
...
+
+            // Emulate vote with ballot masks.
+            // We do that when the GPU thread count is not 32,
+            // since the shader code assumes it is 32.
+            // allInvocations => ballot(pred) == ballot(true),
+            // anyInvocation => ballot(pred) != 0,
+            // allInvocationsEqual => ballot(pred) == balot(true) || ballot(pred) == 0
+            Operand ballotMask = EmitBallot(context, pred);
+
+            Operand AllTrue() => context.ICompareEqual(ballotMask, EmitBallot(context, Const(IrConsts.True)));
```
[the remark this answers is not in the document]
Comment (gdkchan) [reply src/Ryujinx.Graphics.Shader/Instructions/InstEmitWarp.cs:93]: It is used in two places in the switch statement, `VoteMode.All` and `VoteMode.Eq`.

[the remark this answers is not in the document]
Comment (riperiperi) [reply src/Ryujinx.Graphics.Shader/Instructions/InstEmitWarp.cs:93]: Ah, didn't see it in there.

**Review by marysaka — APPROVED** [review]

**Comment by TANTRUM01** [comment]
Just checked with Master Detective, actually works, thank you so much! (Although i've finished it already :D)

**Pull request merged**